lib/src/matchers.rs

```diff
@@ -269,54 +269,67 @@ impl Matcher for IntersectionMatcher<'_> {
 /// visited.
 #[derive(PartialEq, Eq, Debug)]
 struct Dirs {
-    dirs: HashMap<RepoPath, HashSet<RepoPathComponent>>,
-    files: HashMap<RepoPath, HashSet<RepoPathComponent>>,
+    entries: HashMap<RepoPathComponent, Dirs>,
+    // is_dir/is_file aren't exclusive, both can be set to true. If entries is not empty,
+    // is_dir should be set.
+    is_dir: bool,
+    is_file: bool,
 }
 
 impl Dirs {
     fn new() -> Self {
         Dirs {
-            dirs: HashMap::new(),
-            files: HashMap::new(),
+            entries: HashMap::new(),
+            is_dir: false,
```

Comment:
`is_dir` could be encoded that way, but I'm not sure if I like it. Since it's not so clear that `entries.is_some()` means `is_dir`, we'll anyway need inline comment or `is_dir()` function.

And I think `HashMap::new()` is cheap. It doesn't do heap allocation at least.

Reply:
Sounds good.